core/services/ocr2/plugins/ocr2keeper/coordinator.go

```diff
@@ -0,0 +1,158 @@
+	var err error
+
+	contract, err := registry.NewKeeperRegistry(common.HexToAddress("0x"), client)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add log filters for the log poller so that it can poll and find the logs that
+	// we need. Not unregistering the filter later so we ignore the id
+	_, err = logPoller.RegisterFilter(logpoller.Filter{
+		EventSigs: []common.Hash{
+			registry.KeeperRegistryUpkeepPerformed{}.Topic(),
+		},
+		Addresses: []common.Address{registryAddress},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &LogCoordinator{
+		logger:          logger,
+		logPoller:       logPoller,
+		registryAddress: registryAddress,
+		lookbackBlocks:  lookbackBlocks,
+		registry:        contract,
+	}, nil
+}
+
+func (c *LogCoordinator) PerformLogs(ctx context.Context) ([]plugintypes.PerformLog, error) {
+	end, err := c.logPoller.LatestBlock(pg.WithParentCtx(ctx))
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to get latest block from log poller", err)
+	}
+
+	// always check the last lookback number of blocks and rebroadcast
+	// this allows the plugin to make decisions based on event confirmations
+	logs, err := c.logPoller.LogsWithSigs(
+		end-c.lookbackBlocks,
+		end,
+		[]common.Hash{
+			registry.KeeperRegistryUpkeepPerformed{}.Topic(),
+		},
+		c.registryAddress,
+		pg.WithParentCtx(ctx),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to collect logs from log poller", err)
+	}
+
+	performed, err := c.unmarshalLogs(logs)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to unmarshal logs", err)
+	}
+
+	vals := []plugintypes.PerformLog{}
+	for _, p := range performed {
+		if p.BlockNumber < int64(p.CheckBlockNumber) {
+			c.logger.Infof("log block number '%d' is before the block the upkeep was checked at '%d'; not including log in output", p.BlockNumber, p.CheckBlockNumber)
+			continue
+		}
+
+		// broadcast log to subscribers
+		l := plugintypes.PerformLog{
+			Key:           pluginutils.BlockAndIdToKey(big.NewInt(int64(p.CheckBlockNumber)), p.Id),
+			TransmitBlock: plugintypes.BlockKey([]byte(fmt.Sprintf("%d", p.BlockNumber))),
+			Confirmations: p.Confirmations,
+		}
+		vals = append(vals, l)
+	}
+
+	return vals, nil
+}
+
+func (c *LogCoordinator) unmarshalLogs(logs []logpoller.Log) ([]performed, error) {
+	results := []performed{}
+
+	for _, log := range logs {
+		rawLog := toGethLog(log)
+		abilog, err := c.registry.ParseLog(rawLog)
+		if err != nil {
+			return results, err
+		}
+
+		switch l := abilog.(type) {
+		case *registry.KeeperRegistryUpkeepPerformed:
+			if l == nil {
+				continue
+			}
+
+			r := performed{
+				Log:                           log,
+				KeeperRegistryUpkeepPerformed: *l,
+				Confirmations:                 0,
+			}
+
+			results = append(results, r)
+		}
+	}
+
+	return results, nil
+}
+
+type performed struct {
+	logpoller.Log
+	registry.KeeperRegistryUpkeepPerformed
+	Confirmations int64
+}
+
+func toGethLog(lg logpoller.Log) types.Log {
```

Comment:
maybe just put this as a method on logpoller.Log? Can also make use of the existing GetTopics